eras/byron/ledger/executable-spec/src/Byron/Spec/Ledger/Update.hs

```diff
@@ -232,9 +233,13 @@ data UProp = UProp
   }
   deriving (Eq, Generic, Show, Hashable, Data, Typeable, NoThunks)
 
+#if MIN_VERSION_hashable(1,3,4)
+-- Instance Hashable (Set a) is provided by the hashable package
+#else
 -- We need the Hashable instance before making lenses.
 instance Hashable a => Hashable (Set a) where
   hashWithSalt = H.hashUsing Set.toList
+#endif
```

Comment:
Please pay special attention to the change in hash function to ensure this is safe to do.

Reply:
I doubt that this change will cause problems, but I think we need to investigate it in a bit more detail. In particular someone who is familiar with how Hashable instances are being used in byron. This is important because hash value produced by the new hashable version is different. For example if it is used with `unordered-containers` then the order in which values are stored in the HashSet/HasMap will be different and it will affect things like folding and conversion to list:

* `hashable-1.3.4.1`
```haskell
Prelude> import Data.HashSet as HashSet
Prelude HashSet> HashSet.fromList ["foo", "bar", "baz"]
fromList ["bar","foo","baz"]
````
* `hashable-1.3.0.0`
```haskell
Prelude> import Data.HashSet as HashSet
Prelude HashSet> HashSet.fromList ["foo", "bar", "baz"]
fromList ["baz","foo","bar"]
```

I used `unordered-containers` just as an example, I don't think it is used directly in byron. Updating to the new `hashable` version does already change the outcome of `hashWithSalt` function for many types, so that in itself is worth investigating. The only thing about the particular change in this PR that I can find interesting is that hash of a set will not match the one of a matching list (i.e. `xs == Set.toList (Set.fromList xs)) ==> hash (Set.fromList xs) /= hash xs`), which is probably not a big of a deal at all considering that hash values for other types changed as well.

Here is the new version:

* `hashable-1.3.4.1`
```haskell
Prelude> import Data.Set as Set
Prelude Set> import Data.Hashable
Prelude Set Data.Hashable> let s = Set.fromList [1,2,3,4] :: Set Int
Prelude Set Data.Hashable> hashWithSalt (-3750763034362895579) s
6642510073022299387
Prelude Set Data.Hashable> let xs = [1,2,3,4] :: [Int]
Prelude Set Data.Hashable> hashWithSalt (-3750763034362895579) xs
3708964778940489643
```
Note that hash values for list and set differ. 

While this is the old behavior:
* `hashable-1.3.0.0`
```haskell
Prelude> import Data.Set as Set
Prelude Set> import Data.Hashable
Prelude Set Data.Hashable> let s = Set.fromList [1,2,3,4] :: Set Int
Prelude Set Data.Hashable> hashWithSalt (-3750763034362895579) s

<interactive>:4:1: error:
    • No instance for (Hashable (Set Int))
        arising from a use of ‘hashWithSalt’
    • In the expression: hashWithSalt (- 3750763034362895579) s
      In an equation for ‘it’:
          it = hashWithSalt (- 3750763034362895579) s
Prelude Set Data.Hashable> hashUsing Set.toList (-3750763034362895579) s
-6108972173346416629
Prelude Set Data.Hashable> let xs = [1,2,3,4] :: [Int]
Prelude Set Data.Hashable> hashWithSalt (-3750763034362895579) xs
-6108972173346416629
````
Naturally hash values for list and set are the same.

Reply:
I had the same comment. Since this is for Byron, I guess we can get away with just testing it out. If it syncs mainnet, that's all we really care about.

I'm pretty certain the hashes aren't used in anywhere important.